Clase6 tarea wip

Horacaz opened the pull request.

Horacaz committed:
Arreglo boton reincio

Incluyo boton para reinicar los campos y con el fin de realizar nuevamente el ingreso de los datos del grupo familiar.

Horacaz committed:
finalizacion tarea 6

Completadas ambas partes de la tarea 6.

Horacaz committed:
Organizacion del codigo

fsodano commented on clase-6.css:
```diff
@@ -0,0 +1,3 @@
+.oculto {
+    display: none;
+}
```
el simbolito rojo significa que deberías tener una línea vacía al final del archivo, podés configurar vscode para que lo haga automáticamente por vos

fsodano commented on index.html:
```diff
@@ -1,16 +1,50 @@
+    <div>
+        <form>
+            <label for="cantidad-integrantes">Ingresá la cantidad de integrantes de tu grupo familiar:</label>
+            <input type=number id="cantidad-integrantes">
+            <input type=button value="Ingresar" id="ingresar-integrantes">
```
ojo con los "" faltantes en los valores de los atributos

fsodano commented on tarea-clase-6-calculos.js:
```diff
@@ -0,0 +1,71 @@
+function calcularMayorEdad(edadIntegrantes) {
+    let mayorIntegrante = Number(edadIntegrantes[0].value)
+
+    for (i = 0; i < edadIntegrantes.length; i++) {
+        if (mayorIntegrante < Number(edadIntegrantes[i].value)) {
+            mayorIntegrante = Number(edadIntegrantes[i].value)
+        }
+    }
+    return mayorIntegrante;
```
fijate que estás haciendo el cast a Number 3 veces. Esta función debería ya estar esperando números, no hace falta hacer un cast.

fsodano commented on tarea-clase-6-calculos.js:
```diff
@@ -0,0 +1,71 @@
+}
+
+function calcularMenorEdad(edadIntegrantes) {
+    let menorIntegrante = Number(edadIntegrantes[0].value)
+
+    for (i = 0; i < edadIntegrantes.length; i++) {
+        if (menorIntegrante > Number(edadIntegrantes[i].value)) {
+            menorIntegrante = Number(edadIntegrantes[i].value)
+
+        }
```
Al igual que arriba, no hace falta hacer cast, quien llame a calcularMenorEdad debería pasar un listado de números

fsodano commented on tarea-clase-6-calculos.js:
```diff
@@ -0,0 +1,71 @@
+    return (totalEdades / edadIntegrantes.length).toFixed();
+}
+
+function calcularMayorSalarioAnual(salarioAnual) {
```
salarioAnual (singular) es un array de cosas? no tiene sentido, debería ser "salariosAnuales"

fsodano commented on tarea-clase-6-calculos.js:
```diff
@@ -0,0 +1,71 @@
+    mayorSalarioAnual = 0;
+
+    for (i = 0; i < salarioAnual.length; i++) {
+        if (mayorSalarioAnual < Number(salarioAnual[i])){
+            mayorSalarioAnual = Number(salarioAnual[i])
+        }
```
hacé el cast sólo una vez

fsodano commented on tarea-clase-6-calculos.js:
```diff
@@ -0,0 +1,71 @@
+    for (i = 0; i < edadIntegrantes.length; i++) {
+        totalEdades += Number(edadIntegrantes[i].value);
+    }
+    return (totalEdades / edadIntegrantes.length).toFixed();
```
La función que haya llamado a calcularPromedio es la que tiene que lidiar con decidir cuántos decimales mostrar

fsodano commented on tarea-clase-6-calculos.js:
```diff
@@ -0,0 +1,71 @@
+    let totalEdades = 0
+
+    for (i = 0; i < edadIntegrantes.length; i++) {
+        totalEdades += Number(edadIntegrantes[i].value);
```
mismo comentario que arriba, acá estás tomando un listado de elementos, que está mal. Estas funciones sólo deberían lidiar con números

fsodano commented on tarea-clase-6-calculos.js:
```diff
@@ -0,0 +1,71 @@
+    return mayorSalarioAnual;
+}
+
+function calcularMenorSalarioAnual(salarioAnual) {
+    menorSalarioAnual = Number(salarioAnual[0]);
+
+    for (i = 0; i < salarioAnual.length; i++) {
+        if (menorSalarioAnual > Number(salarioAnual[i])){
+            menorSalarioAnual = Number(salarioAnual[i])
+        }
+    }
+    return menorSalarioAnual
+}
+
+function calcularSalarioMensualPromedio(salarioAnual) {
+    salarioMensualPromedio = 0;
+
+    for (i = 0; i < salarioAnual.length; i++) {
+        salarioMensualPromedio += Number(salarioAnual[i]);
+    }
+    return (salarioMensualPromedio / 12).toFixed(2)
+}
+
+function calcularSalarioAnualPromedio(salarioAnual) {
+    salarioAnualPromedio = 0;
+    for (i = 0; i < salarioAnual.length; i++) {
+        salarioAnualPromedio += Number(salarioAnual[i]);
+    }
+    return (salarioAnualPromedio / salarioAnual.length).toFixed(2)
+}
```
Estás repitiendo siempre el mismo algoritmo, deberías tener 3 funciones genéricas:

calcularMayor
calcularMenor
calcularPromedio

y reusarlas.

fsodano commented on tarea-clase-6.js:
```diff
@@ -6,6 +6,127 @@ Al hacer click en "calcular", mostrar en un elemento pre-existente la mayor edad
+
+$BOTON_CANTIDAD_INTEGRANTES.onclick = function () {
+
+    cantidadIntegrantes = document.querySelector("#cantidad-integrantes").value;
```
esta variable no está declarada en ningún lado por lo tanto se convierte en global que te puede traer efectos inesperados

fsodano commented on tarea-clase-6.js:
```diff
@@ -6,6 +6,127 @@ Al hacer click en "calcular", mostrar en un elemento pre-existente la mayor edad
+
+    cantidadIntegrantes = document.querySelector("#cantidad-integrantes").value;
+
+    if (cantidadIntegrantes == 0) {
+        return alert("Ingresá un valor mayor a 0");
+    }
```
y si pongo `-1`?

fsodano commented on tarea-clase-6.js:
```diff
@@ -6,6 +6,127 @@ Al hacer click en "calcular", mostrar en un elemento pre-existente la mayor edad
+    }
+
+    for (i = 0; i < cantidadIntegrantes; i++) {
+        imprimirCantidadIntegrantes(i);
```
crearIntegrantes me parece un mejor nombre, porque no estás "imprimiendo" nada

fsodano commented on tarea-clase-6.js:
```diff
@@ -6,6 +6,127 @@ Al hacer click en "calcular", mostrar en un elemento pre-existente la mayor edad
+
+}
+
+function mostrarBotonCalcularEdad() {
+    document.querySelector("#calculo-edad").className = "";
+
+}
```
quizás sea mejor tener 2 funciones genéricas:


mostrarElemento
ocultarElemento

que sepan agregar y quitar la clase "oculto"

fijate que esta función lo que es limpiar TODAS las clases que quizás no es lo que quieras en un futuro.

fsodano commented on tarea-clase-6.js:
```diff
@@ -6,6 +6,127 @@ Al hacer click en "calcular", mostrar en un elemento pre-existente la mayor edad
+    return false;
+}
+
+function mostrarMensajeMayorEdad(calcularMayorEdad) {
```
calcularMayorEdad por tener su primera parte como un verbo en infinitivo parece una función, en este caso el parámetro debería ser "mayorEdad" para dar a entender que es una variable que contiene la mayor edad.

fsodano commented on tarea-clase-6.js:
```diff
@@ -6,6 +6,127 @@ Al hacer click en "calcular", mostrar en un elemento pre-existente la mayor edad
+    const integrantes = document.querySelectorAll(".integrante");
+
+    for (i = 0; i < integrantes.length; i++) {
+        const integrante = document.querySelector(".integrante");
```
ya tenés toda la lista de integrantes en la variable `integrantes`, no te hace falta hacer otro querySelector

fsodano reviewed (changes requested):
Te dejé varios comentarios, los puntos más importantes son:

* Reusá más código (ahora hiciste 269 líneas, fijate cuántas líneas te quedan cuando hagas las correcciones)
* No pases elementos de HTML para hacer algoritmos, pasá números
* Arreglá el nombre de los parámetros para no confundir funciones con variables. Funciones: hacerAlgo, variables: algo

Horacaz committed:
Corrección calculos

se refactorizan los calculos con el fin de ser reutilizados en todo el codigo, eliminando la repeticion innecesaria de de algoritmos similares.

Horacaz committed:
correccion index

Se modifica el index para agregar los mensajes que muestran los calculos de los datos introducidos por el usuario. Estos mensajes antes eran mostrados por el script, pero ya que los mensajes a mostrar ya se encuentran predefinidos se opta por incluirlos directamente en el index, modificando su valor a través del script, mejorando la lectura del código.

Horacaz committed:
Correccion tarea-clase-6.js

Se modifica el codigo teniendo en cuenta los comentarios del profesor, reduciendo lineas de código, reutilizando funciones y algoritmos.

Horacaz committed:
Agrega linea vacía clase-6.css
